framework/test_app/runners/k8s/k8s_base_runner.py

```diff
@@ -74,17 +81,45 @@ class KubernetesBaseRunner(base_runner.BaseRunner, metaclass=ABCMeta):
     # Fields with default values.
     namespace_template: str = "namespace.yaml"
     reuse_namespace: bool = False
+    log_to_stdout: bool = False
 
-    # Mutable state. Describes the current run.
+    # The history of all runs performed by this runner.
+    # Persisted across multiple runs.
+    run_history: collections.deque[RunHistory]
+
+    # Below is mutable state associated with the current run.
     namespace: Optional[k8s.V1Namespace] = None
     deployment: Optional[k8s.V1Deployment] = None
     deployment_id: Optional[str] = None
     service_account: Optional[k8s.V1ServiceAccount] = None
+
+    # A map of pod names to pod objects as they were at the moment
+    # of deployment creation.
+    # All user-requested operations that modify the deployment directly (such as
+    # scaling, manual pod deletion, etc.) must update this mapping.
+    # Note that this doesn't represent the current state of the pods, so
+    # don't update pod definitions just because they loaded at a later time.
+    pods_started: dict[str, k8s.V1Pod]
+    # When a pod is being shut down due to a user-requested operation
+    # described above, delete it from pods_created, and move to pods_stopping.
+    pods_stopping: dict[str, k8s.V1Pod]
```

Comment:
Do we allow multi-threaded access to this class? If so, we'll probably want some synchronization on these. If not, we'll probably want to explicitly document that.

Reply:
No, no multithreading/async that would mutate these methods. Thanks for pointing this out. Great idea to document this!

Reply:
added a note to the class docstring